### ISSUE TITLE
How can we train LSTM PPO?

### ISSUE THREAD

**Issue by jd730** [astooke/rlpyt#85]
Hi,
I am trying to train LSTM PPO on Hopper-v3 but it did not learn well.
Although LSTM policy is hard to learn compared to FF policy, it seems there are several missing pieces to train them.

Could you give some advice to train LSTM PPO?

Thank you so much.

![image](https://user-images.githubusercontent.com/9247850/70881320-75224580-200f-11ea-915f-e1322ff5dbfe.png)

cf. my settings are as follows:

```
algo=dict(
        discount=0.99,
        learning_rate=3e-4,
        clip_grad_norm=1e6,
        entropy_loss_coeff=0.0,
        gae_lambda=0.95,
        minibatches=128,
        epochs=10,
        ratio_clip=0.2,
        normalize_advantage=True,
        linear_lr_schedule=True,
        bootstrap_timelimit=False,
    ),
    sampler=dict(
        batch_B=16,
        max_decorrelation_steps=400,
    ),
```
I tried batch_T as 2048, 256, 40, 32, 16

cf2. LSTM did not increase after finishing point.

**Comment by tarungog** [comment]
@jd730  what is this nice vis tool you are using?

**Comment by jd730** [comment]
@tarungog  Hi, I am using [wandb](https://www.wandb.com/).

**Comment by astooke** [comment]
Hi!  Looking at your settings, I think `minibatches` in the algo might be too high relative to the sampler `batch_B`.   Try using `batch_B` as different multiples of `minibatches`, and let us know if that works?

`minibatches=128` will try to chop the training batch into 128 minibatches.  But in PPO with recurrence, the shuffling/minibatching all happens along the B dimension only, so the agent receives full-length sequences in T.  I'm not sure what the behavior will be when `minibatches` > `batch_B`....should add a warning about this!  In feedforward mode, the shuffling/minibatching happens over individual data points (T and B).

**Comment by jd730** [comment]
@astooke  Thank you for your advice.
I tested various settings as follows:
set LSTM size as fc (64) - lstm(64) since FF default setting is fc(64)-fc(64).
**(minibatch, batch_T, batch_B)**: (4, 20, 16), (4, 40, 16), (4, 40, 48), (16, 40, 48), (16, 256, 16), (16, 40, 16), (32, 40, 48).

![image](https://user-images.githubusercontent.com/9247850/71329772-14918c00-256c-11ea-8654-1cfc2784b17e.png)

The best setting (blue) was (32,40, 48) setting but decreased after 40M and much lower than FF setting (brown).

Furthermore, in LSTM, I am figuring out that sometimes entropy goes to infinity. it randomly happens. i.e. if I ran three jobs with the same settings, 0-1 job works like below.

![image](https://user-images.githubusercontent.com/9247850/71329866-b6b17400-256c-11ea-9d81-1b12b22f1d18.png)

**Comment by astooke** [comment]
Hi!  I'm not sure about the specifics of this learning problem, so I might not be able to help much.  But one other thing to try is to put a skip connection around the LSTM, and maybe keep 2 FF layers before it, so your network would like like fc1-->fc2-->lstm-->(+)fc2_output-->output.  So the network can learn as FF when that's OK, and just pick up recurrence where it's helpful later on in training.

Let us know what works!

**Comment by jd730** [comment]
@astooke, Thanks a lot. I will try following your advice and update it.

**Comment by MasterScrat** [comment]
@jd730 off-topic, but would you mind sharing your wandb integration code?

**Comment by astooke** [comment]
Hi!  @jd730 any progress here?  Else I would like to close the issue, as it's about a specific RL problem rather than the code base.  Thanks! :)

**Comment by jd730** [comment]
Hi, @astooke. Unfortunately, due to other works, I did not try yet. I'll close it and if I have any updates, I will reopen and share the results. 

Thank you for your help.

@MasterScrat, you may add some lines regarding wandb on `dump_tabular` on `logger.py`

**Comment by yannidd** [comment]
TLDR: I think I found a solution: set `value_loss_coeff=1e-3` and `batch_B=minibatches`.

This issue is now closed but it seems that no solution was found; I was recently training a MujocoLstmAgent on a custom robot environment and initially had the same problem of the agent not being able to learn. After a quick inspection, I saw that the actor and the critic are implemented in a single network; the following call `print(agent.model)` yields:

```
MujocoLstmModel(
  (mlp): MlpModel(
    (model): Sequential(
      (0): Linear(in_features=77, out_features=256, bias=True)
      (1): ReLU()
      (2): Linear(in_features=256, out_features=256, bias=True)
      (3): ReLU()
    )
  )
  (lstm): LSTM(275, 256)
  (head): Linear(in_features=256, out_features=37, bias=True)
)
```

According to a similar issue (ray-project/ray#5278), the problem comes from the fact that the value function loss is huge when compared to the policy loss. The solution is to set the `value_loss_coeff` parameter to a low value (I used 1e-3) to balance the two losses. Below is the full configuration I used (I normally set `batch_B=minibatches`):

```python
  sampler = Sampler(
      ...
      EnvCls=gym_make,
      CollectorCls=CpuWaitResetCollector,
      batch_T=256,
      batch_B=8,
      max_decorrelation_steps=0,
  )
  algo = PPO(
      ...
      discount=0.99,
      learning_rate=3e-4,
      clip_grad_norm=1e6,
      entropy_loss_coeff=0.0,
      gae_lambda=0.95,
      minibatches=8,
      epochs=10,
      value_loss_coeff=1e-3,
      ratio_clip=0.2,
      normalize_advantage=True,
      linear_lr_schedule=False,
  )
  agent = MujocoLstmAgent(model_kwargs=dict(normalize_observation=False))
```

Hope this helps someone :smile:
